### PR TITLE
Add logic key for tileable graph

### PR DESCRIPTION
**Pull request by zhongchun**
## What do these changes do?
 Logic key is a unique and deterministic key for `TileableGraph`. For multiple runs the logic key will remain same if the computational logic doesn't change. And it can be used to some optimization when running a same `execute`, like HBO.

## Related issue number
#2787 

## Check code requirements

- [ ] tests added / passed (if needed)
- [ ] Ensure all linting tests pass, see [here](https://docs.pymars.org/en/latest/development/contributing.html#check-code-styles) for how to run them

### CODE REVIEW

File: mars/core/graph/entity.py [hunk 1]
```diff
@@ -74,6 +84,20 @@ def results(self):
...
+    def logic_key(self):
+        if not hasattr(self, "_logic_key") or self._logic_key is None:
+            token_keys = []
+            for node in self.topological_iter():
+                token_keys.append(
+                    tokenize(node.op.get_logic_key(), **node.extra_params)
+                )
+            self._logic_key = tokenize(*sorted(token_keys))
+            logger.debug(
+                "Got logic key: %s, token_keys: %s", self._logic_key, token_keys
```
Comment (wjsi) [inline mars/core/graph/entity.py:97]: May just need to dump keys for tasks. Keys for all nodes bring too much information which may not be helpful for debug.

File: mars/core/graph/entity.py [hunk 2]
```diff
@@ -74,6 +84,20 @@ def results(self):
     def results(self, new_results):
         self._result_tileables = new_results
 
+    @property
+    def logic_key(self):
+        if not hasattr(self, "_logic_key") or self._logic_key is None:
+            token_keys = []
+            for node in self.topological_iter():
```
Comment (wjsi) [inline mars/core/graph/entity.py:91]: `topological_iter` does not guarantee order for nodes.

File: mars/core/graph/entity.py [hunk 3]
```diff
@@ -74,6 +84,20 @@ def results(self):
...
+                token_keys.append(
+                    tokenize(node.op.get_logic_key(), **node.extra_params)
```
Comment (wjsi) [inline mars/core/graph/entity.py:93]: As keys already determinded for nodes, is tehre any extra information brought by extra_params? Remember that `tokenize` is an expensive call.

File: mars/core/graph/entity.py (hunk 2 above)
Comment (zhongchun) [reply mars/core/graph/entity.py:91]: Year, so sort the `token_keys` in line 93.

File: mars/core/graph/entity.py (hunk 3 above)
Comment (zhongchun) [reply mars/core/graph/entity.py:93]: Thanks for your reminder. There are `raw_chunk_size` and other info in extra_params which may make the chunk graph is different. This logic key is used when communicating with external service and its cost is comparable to the time consumed by requesting external service.

Comment (wjsi) [reply mars/core/graph/entity.py:93]: I would like `tokenize` not performend when there is no `extra_params`.

File: mars/core/graph/entity.py (hunk 2 above)
Comment (wjsi) [reply mars/core/graph/entity.py:91]: Then `topological_iter` can even be replaced with a BFS to save time.

File: mars/core/graph/entity.py (hunk 3 above)
Comment (zhongchun) [reply mars/core/graph/entity.py:93]: Good suggestion.

File: mars/core/graph/entity.py (hunk 2 above)
Comment (zhongchun) [reply mars/core/graph/entity.py:91]: Good suggestion. Thanks a lot.

File: mars/core/graph/tests/test_graph.py [hunk 4]
```diff
@@ -105,3 +108,75 @@ def test_to_dot():
...
     dot = to_str(graph.to_dot(trunc_key=5))
     assert all(to_str(n.key)[:5] in dot for n in graph) is True
+
+
+def test_tileable_graph_logic_key():
+    # Tensor
+    t1 = mt.random.randint(10, size=(10, 8), chunk_size=4)
+    t2 = mt.random.randint(10, size=(10, 8), chunk_size=5)
+    graph1 = (t1 + t2).build_graph(tile=False)
+    graph2 = (t1 + t2).build_graph(tile=False)
```
Comment (fyrestone) [inline mars/core/graph/tests/test_graph.py:118]: Is the logic key remains the same even the tileable instances are different?
```python
graph1 = (t1 + t2).build_graph(tile=False)
tt1 = mt.random.randint(10, size=(10, 8), chunk_size=4)
tt2 = mt.random.randint(10, size=(10, 8), chunk_size=5)
graph2 = (tt1 + tt2).build_graph(tile=False)
assert graph1.logic_key == graph2.logic_key
```

Comment (zhongchun) [reply mars/core/graph/tests/test_graph.py:118]: Yes. Your writing is better, I'll optimize this.

Comment (fyrestone) [reply mars/core/graph/tests/test_graph.py:118]: Good to know this.

**Review by wjsi — APPROVED** [review]
LGTM

**Review by qinxuye — APPROVED** [review]
LGTM

**Pull request merged**